src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -3185,6 +3186,39 @@ void c_typecheck_baset::typecheck_expr_binary_arithmetic(exprt &expr)
 
   // promote!
 
+  // warn about an implicit 0-extension of bitand operands
+  // expect an explicit cast to avoid this warning
+  if((expr.id()==ID_bitand) &&
+     (op0.type().id()==ID_signedbv || op0.type().id()==ID_unsignedbv) &&
+     (op1.type().id()==ID_signedbv || op1.type().id()==ID_unsignedbv))
+  {
+    bool unintentional=false;
+    // do not apply, if we are working with constants and the & operator
+    if(op0.id()!=ID_constant && op1.id()!=ID_constant)
+    {
+      // mismatch occurs only, if we have "&" with two different width,
+      // where the side with the smaller width has to use the "~" operator
+      const size_t width0=to_bitvector_type(op0.type()).get_width();
+      const size_t width1=to_bitvector_type(op1.type()).get_width();
+      unintentional=(width0<width1 && has_subexpr(op0, ID_bitnot)) ||
+                    (width0>width1 && has_subexpr(op1, ID_bitnot));
+    }
+
+    // is all conditions are met for an unintended combination
+    // of the operators ~ and &, print a warning
+    if(unintentional)
+    {
+      const source_locationt &location=expr.find_source_location();
+      warning() << concat_dir_file(id2string(location.get_working_directory()),
```

Comment:
I'll use the proper CBMC way. as discussed offline already, the output format of warnings and errors does not match the format of gcc.